core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,317 @@
+    }
+    val leaderIsrAndControllerEpochs = (partitionsWhoseIsrExcludesReplica ++ successfulUpdates).map { case (partition, leaderAndIsr) =>
+      val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controller.epoch)
+      controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+      partition -> leaderIsrAndControllerEpoch
+    }
+    (leaderIsrAndControllerEpochs, updatesToRetry)
+  }
+
+  // replicas not in isr are implied as partitions in currentLeaderAndIsrs that are not in candidateLeaderAndIsrs
+  private def getTopicPartitionStatesFromZk(replicaId: Int, partitions: Seq[TopicAndPartition]):
+  (Map[TopicAndPartition, LeaderAndIsr],
+    Map[TopicAndPartition, LeaderAndIsr],
+    Seq[TopicAndPartition],
+    Map[TopicAndPartition, Exception]) = {
+    val currentLeaderAndIsrs = mutable.Map.empty[TopicAndPartition, LeaderAndIsr]
+    val candidateLeaderAndIsrs = mutable.Map.empty[TopicAndPartition, LeaderAndIsr]
+    val partitionsWithNoLeaderAndIsrInZk = mutable.Buffer.empty[TopicAndPartition]
+    val failed = mutable.Map.empty[TopicAndPartition, Exception]
+    val getDataResponses = try {
+      KafkaControllerZkUtils.getTopicPartitionStates(zookeeperClient, partitions)
+    } catch {
+      case e: Exception =>
+        partitions.foreach(partition => failed.put(partition, e))
+        return (currentLeaderAndIsrs.toMap, candidateLeaderAndIsrs.toMap, partitionsWithNoLeaderAndIsrInZk, failed.toMap)
+    }
+    getDataResponses.foreach { getDataResponse =>
+      val partition = getDataResponse.ctx.asInstanceOf[TopicAndPartition]
+      if (Code.get(getDataResponse.rc) == Code.OK) {
+        val leaderIsrAndControllerEpochOpt = TopicPartitionStateZNode.decode(getDataResponse.data, getDataResponse.stat)
+        if (leaderIsrAndControllerEpochOpt.isEmpty) {
+          partitionsWithNoLeaderAndIsrInZk += partition
+        } else {
+          val leaderIsrAndControllerEpoch = leaderIsrAndControllerEpochOpt.get
+          if (leaderIsrAndControllerEpoch.controllerEpoch > controller.epoch) {
+            val exception = new StateChangeFailedException("Leader and isr path written by another controller. This probably" +
+              s"means the current controller with epoch ${controller.epoch} went through a soft failure and another " +
+              s"controller was elected with epoch ${leaderIsrAndControllerEpoch.controllerEpoch}. Aborting state change by this controller")
+            failed.put(partition, exception)
+          } else {
+            currentLeaderAndIsrs.put(partition, leaderIsrAndControllerEpoch.leaderAndIsr)
+            if (leaderIsrAndControllerEpoch.leaderAndIsr.isr.contains(replicaId)) {
+              val newLeader = if (replicaId == leaderIsrAndControllerEpoch.leaderAndIsr.leader) LeaderAndIsr.NoLeader else leaderIsrAndControllerEpoch.leaderAndIsr.leader
+              val candidateIsr = leaderIsrAndControllerEpoch.leaderAndIsr.isr.filter(_ != replicaId)
```

Comment:
The method name is a bit confusing. It sounds like that we are just reading the partition state from ZK, but we actually also remove the isr here. So, we probably want to use a more accurate method name. Also, would it be better to change the isr in leaderAndIsrBasedOnLogConfigs() instead of here?